packages/firestore/test/unit/specs/spec_test_runner.ts

```diff
@@ -793,13 +813,9 @@ abstract class TestRunner {
 
   private doWriteAck(writeAck: SpecWriteAck): Promise<void> {
     const updateTime = this.serializer.toVersion(version(writeAck.version));
-    const nextMutation = this.outstandingMutations.poll();
+    const nextMutation = this.sharedWrites.shift();
     return this.validateNextWriteRequest(nextMutation).then(() => {
       this.connection.ackWrite(updateTime, [{ updateTime }]);
```

Comment:
FWIW- ackWrite() will end up calling SyncEngine.applySuccessfulWrite() which will end up resolving the Deferred<> representing the user callback, but since Promise continuations are called asynchronously, the actual user callback we created (that updates our acknowledged/rejectedDocs lists) won't run synchronously.  I'm guessing we do enough extra Promise/setTimeout gunk that the callback ends up deterministically running before we check the expectations, but since it's no longer explicit this could be fragile.  I think I'm okay with it... but it could be worth a comment somewhere or something perhaps.  Up to you.

Reply:
The Spec Tests call `this.queue.drain()` before invoking `validateStepExpectations. Since drain uses setTimeout(), it waits for all Promise chains to resolve beforehand.